lib/printer.js

```diff
@@ -962,10 +1017,227 @@ function genericPrintNoParens(path, options, print) {
     case "Line": // Line comment.
         return concat(["//", fromString(n.value, options)]);
 
+    // Type Annotations for Facebook Flow, typically stripped out or
+    // transformed away before printing.
+    case "TypeAnnotation":
+        return concat([
+            n.typeAnnotation.type != "FunctionTypeAnnotation" ? ": " : "",
+            path.call(print, "typeAnnotation")
+        ]);
+
+    case "AnyTypeAnnotation":
+        return fromString("any", options);
+
+    case "ArrayTypeAnnotation":
+        return concat([
+            path.call(print, "elementType"),
+            "[]"
+        ]);
+
+    case "BooleanTypeAnnotation":
+        return fromString("boolean", options);
+
+    case "DeclareClass":
+        return concat([
+            fromString("declare class ", options),
+            path.call(print, "id"),
+            " ",
+            path.call(print, "body"),
+        ]);
+
+    case "DeclareFunction":
+        return concat([
+            fromString("declare function ", options),
+            path.call(print, "id")
+        ]);
+
+    case "DeclareModule":
+        return concat([
+            fromString("declare module ", options),
+            path.call(print, "id"),
+            " ",
+            path.call(print, "body"),
+        ]);
+
+    case "DeclareVariable":
+        return concat([
+            fromString("declare var ", options),
+            path.call(print, "id")
+        ]);
+
+    case "FunctionTypeAnnotation":
+        // yay magic. FunctionTypeAnnotation is ambigous:
+        // declare function(a: B): void; OR
+        // var A: (a: B) => void;
```

Comment:
This is the only super-clowntown part of this implementation. It turns out that `FunctionTypeAnnotation` is ambiguous and renders differently based on the context around it.